### PR TITLE
Fixes tunnel startup options handling

### PR DESCRIPTION
**Pull request by diegoreymendez**
Task/Issue URL: https://app.asana.com/0/0/1205466155643247/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2003
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1581
What kind of version bump will this require?: Patch

## Description

Fixes tunnel startup options handling, and improves the options loading mechanism to make it unit testable.

## Steps to test this PR

Follow the testing steps in the macOS and iOS PRs.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/NetworkProtection/ExtensionMessage/ExtensionMessage.swift [hunk 1]
```diff
@@ -65,7 +65,10 @@ public enum ExtensionMessage: RawRepresentable {
         case .isHavingConnectivityIssues:
             self = .isHavingConnectivityIssues
         case .setSelectedServer:
-            guard data.count > 1 else { return nil }
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/ExtensionMessage/ExtensionMessage.swift:68]: This was off, empty data didn't mean to ignore the message... it was a request to reset the selected server to automatic.

File: Sources/NetworkProtection/NetworkProtectionOptionKey.swift [hunk 2]
```diff
@@ -29,6 +29,3 @@ public enum NetworkProtectionOptionKey {
     public static let excludedRoutes = "excludedRoutes"
     public static let connectionTesterEnabled = "connectionTesterEnabled"
 }
-public enum NetworkProtectionOptionValue {
-    public static let `true` = "true" as NSString
-}
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/NetworkProtectionOptionKey.swift:34]: We can send bools as numbers.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 3]
```diff
@@ -37,12 +37,18 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
         case couldNotGenerateTunnelConfiguration(internalError: Error)
         case couldNotFixConnection
         case simulateTunnelFailureError
 
         var errorDescription: String? {
             switch self {
+            case .startingTunnelWithoutOptions:
+                return "Missing tunnel options at startup"
+            case .startingTunnelWithoutAuthToken:
+                return "Missing auth token at startup"
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:51]: Two new errors for handling specific scenarios at tunnel startup.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 4]
```diff
@@ -184,19 +190,22 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
     }
 
     private func setKeyValidity(_ interval: TimeInterval?) {
-        guard keyValidity != interval,
-            let interval = interval else {
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:188]: A `nil` interval wasn't meant to bail out... it's meant to be a request to reset the key validity interval.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 5]
```diff
@@ -345,29 +349,40 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
-              let keyValidity = TimeInterval(keyValidityString) else {
-            return
+    private func loadKeyValidity(from options: StartupOptions) {
+        switch options.keyValidity {
+        case .set(let validity):
+            setKeyValidity(validity)
+        case .useExisting:
+            break
+        case .reset:
+            setKeyValidity(nil)
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:359]: Now we have a basic mechanism to know whether we need to set the validity, use the existing one or reset it.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 6]
```diff
@@ -345,29 +349,40 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
-        guard let serverName = options[NetworkProtectionOptionKey.selectedServer] as? String else {
-            return
+    private func loadSelectedServer(from options: StartupOptions) {
+        switch options.selectedServer {
+        case .set(let selectedServer):
+            selectedServerStore.selectedServer = selectedServer
+        case .useExisting:
+            break
+        case .reset:
+            selectedServerStore.selectedServer = .automatic
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:370]: Same as above but for server selection.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 7]
```diff
@@ -345,29 +349,40 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
+        switch options.authToken {
+        case .set(let authToken):
+            try tokenStore.store(authToken)
+        case .useExisting:
+            break
+        case .reset:
+            // This case should in theory not be possible, but it's ideal to have this in place
+            // in case an error in the controller on the client side allows it.
+            try tokenStore.deleteToken()
+            throw TunnelError.startingTunnelWithoutAuthToken
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:384]: Same as above but for the auth token.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 8]
```diff
@@ -389,47 +404,50 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
     open override func startTunnel(options: [String: NSObject]?, completionHandler: @escaping (Error?) -> Void) {
         connectionStatus = .connecting
-
-        // when activated by system "on-demand" the option is set
-        var isOnDemand: Bool {
-            options?[NetworkProtectionOptionKey.isOnDemand] as? Bool == true
-        }
-        var isActivatedFromSystemSettings: Bool {
-            options?[NetworkProtectionOptionKey.activationAttemptId] == nil && !isOnDemand
-        }
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:399]: This is now handled within `StartupOptions`.

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 9]
```diff
@@ -389,47 +404,50 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
+        tunnelHealth.isHavingConnectivityIssues = false
+        controllerErrorStore.lastErrorMessage = nil
 
         let internalCompletionHandler = { [weak self] (error: Error?) in
-            if error != nil {
-                self?.connectionStatus = .disconnected
-                completionHandler(error)
+            guard let error else {
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:411]: Reverted this condition using a guard so things moved around a bit... sorry!

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 10]
```diff
@@ -389,55 +426,41 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
...
-        }
-        var isActivatedFromSystemSettings: Bool {
-            options?[NetworkProtectionOptionKey.activationAttemptId] == nil && !isOnDemand
-        }
+        tunnelHealth.isHavingConnectivityIssues = false
+        controllerErrorStore.lastErrorMessage = nil
 
         let internalCompletionHandler = { [weak self] (error: Error?) in
-            if error != nil {
-                self?.connectionStatus = .disconnected
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:403]: Tis was moved to line 422.  The diff in this file is really awful.  Apologies!

File: Sources/NetworkProtection/PacketTunnelProvider.swift [hunk 11]
```diff
@@ -342,32 +369,42 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
         let vendorOptions = provider?.providerConfiguration
 
         loadRoutes(from: vendorOptions)
-        self.isConnectionTesterEnabled = vendorOptions?[NetworkProtectionOptionKey.connectionTesterEnabled] as? Bool ?? true
```
Comment (diegoreymendez) [inline Sources/NetworkProtection/PacketTunnelProvider.swift:345]: This should not have been in vendor options.  In fact it was not disabled in iOS due to this.

I've moved it to line 362 so it's part of the startup options instead.

**Review by graeme — APPROVED** [review]
Looks great! Nice extraction of the options. Maybe have a look at those linter issues but I trust you to figure out whether to fix or whether they’re invalid.

**Pull request merged**